### PR TITLE
Multi-Product Selection and track products quantities

### PR DESCRIPTION
**Pull request by AbdullahAbdelhakeem6484**
Upgraded system for selecting multiple products seamlessly. 
Implemented robust error handling for product availability, quantity checks, and total money validation. Optimized code for improved performance and clear error messages. 

Review and feedback are welcome.

### CODE REVIEW

File: src/VendingMachine.cpp [hunk 1]
```diff
@@ -41,30 +41,59 @@ VMErrorCode VendingMachine::selectProduct(std::string productName)
...
+                    } else {
+                        std::cout << "Not enough money to purchase: " << productName << std::endl;
+                        std::cout << "Total money spent so far: " << getTotalMoney() << std::endl;
+                        
+                        // Refund the user and reset the state
+                        moneyAmount = 0;
+                        state = VMState::IDLE;
+                        selectedProducts.clear();
+                        return VMErrorCode::NOT_ENOUGH_MONEY;
+                    }
```
Comment (MHeshamG) [inline src/VendingMachine.cpp:76]: In this case you need to check first if there is a product selected or not if not then move to has_money state and don't refund and clear the selected list just ignore this item

File: src/VendingMachine.cpp [hunk 2]
```diff
@@ -41,30 +41,59 @@ VMErrorCode VendingMachine::selectProduct(std::string productName)
...
     return VMErrorCode::SUCCESS;
 }
 
+double VendingMachine::getTotalMoneyOfSelectedProducts() const {
+    double totalMoneyOfSelectedProducts = 0;
+    for (const auto& product : selectedProducts) {
+        totalMoneyOfSelectedProducts += product.getPrice();
+    }
+    return totalMoneyOfSelectedProducts;
+}
```
Comment (MHeshamG) [inline src/VendingMachine.cpp:94]: why don't you just use totalMoney varaible used before

File: src/VendingMachine.cpp [hunk 3]
```diff
@@ -76,14 +105,18 @@ VMErrorCode VendingMachine::dispenseProduct()
...
-                if(moneyAmount > 0){
-                    std::cout<<"Returning change: "<<moneyAmount<<std::endl;
+                moneyAmount -= selectedProducts.back().getPrice();
+                if (moneyAmount > 0) {
+                    std::cout << "Returning change: " << moneyAmount << std::endl;
                     moneyAmount = 0;
                 }
-                std::cout<<"Dispensing product: "<<choice.getName()<<std::endl;
+
+                std::cout << "Dispensing product: " << selectedProducts.back().getName() << std::endl;
```
Comment (MHeshamG) [inline src/VendingMachine.cpp:114]: you are just dispensing the last product loop over the whole list

File: src/VendingMachine.cpp [hunk 4]
```diff
@@ -76,14 +105,18 @@ VMErrorCode VendingMachine::dispenseProduct()
             return VMErrorCode::NO_PRODUCT_SELECTED;
         break;
         case VMState::PRODUCT_SELECTED:
-                moneyAmount -= choice.getPrice();
-                if(moneyAmount > 0){
-                    std::cout<<"Returning change: "<<moneyAmount<<std::endl;
+                moneyAmount -= selectedProducts.back().getPrice();
```
Comment (MHeshamG) [inline src/VendingMachine.cpp:108]: deduct the total price not the price of last Item only

File: src/Product.cpp [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 using vendingmachine::Product;
 
-Product::Product(std::string name, double price, std::string details) : name{name}, price{price}, details{details}
+Product::Product(std::string name, double price, std::string details, int quantity) : name{name}, price{price}, details{details}, quantity{quantity}
```
[the remark this answers is not in the document]
Comment (AbdullahAbdelhakeem6484) [reply src/Product.cpp:5]: it is related to product

File: src/VendingMachine.cpp (hunk 1 above)
Comment (AbdullahAbdelhakeem6484) [reply src/VendingMachine.cpp:76]: ok, great

File: src/VendingMachine.cpp (hunk 2 above)
Comment (AbdullahAbdelhakeem6484) [reply src/VendingMachine.cpp:94]: you are right i will use it

**Pull request closed without merging**